packages/wonder-blocks-dropdown/components/dropdown-core-virtualized.js

```diff
@@ -1,113 +1,206 @@
 // @flow
 import * as React from "react";
+import ReactDOM from "react-dom";
 import {VariableSizeList as List} from "react-window";
 
-import type {StyleType} from "@khanacademy/wonder-blocks-core";
-
+import DropdownVirtualizedItem from "./dropdown-core-virtualized-item.js";
+import SearchTextInput from "./search-text-input.js";
 import SeparatorItem from "./separator-item.js";
 
 import type {DropdownItem} from "../util/types.js";
 
-// copied from https://github.com/bvaughn/react-window/blob/master/src/createListComponent.js#L17
-type ItemProps = {|
+import {DROPDOWN_ITEM_HEIGHT, SEARCH_ITEM_HEIGHT} from "../util/constants.js";
+
+type Props = {|
     /**
      * The complete list of items that will be virtualized.
      */
     data: Array<DropdownItem>,
 
     /**
-     * The current item index.
+     * The VariableSizeList ref that needs to be passed to access to
+     * react-window's instance methods
      */
-    index: number,
-
-    /**
-     * Whether the item is scrolling or not.
-     */
-    isScrolling?: boolean,
+    listRef?: {
+        current: null | React.ElementRef<typeof List>,
+    },
+|};
 
+type State = {|
     /**
-     * Custom styles passed from react-window
+     * The list width that needs to be passed in order to let react-window
+     * calculate each item position
      */
-    style: StyleType,
+    width: ?number,
+    height: ?number,
 |};
 
 /**
- * A virtualized list item - It's created by decorating the DropdownItem
- * (ActionItem, OptionItem, SeparatorItem) with custom styles to let
- * react-window make its own calculations.
+ * Maximum visible items inside the dropdown list
  */
-class DropdownVirtualizedItem extends React.Component<ItemProps> {
-    render() {
-        const {data, index, style} = this.props;
-        const item = data[index];
+const MAX_VISIBLE_ITEMS = 10;
 
-        if (SeparatorItem.isClassOf(item.component)) {
-            // add react-window style to the separator to preserve the correct
-            // position
-            return React.cloneElement(item.component, {style});
-        } else {
-            const {component, populatedProps, onClick, role, ref} = item;
-
-            return React.cloneElement(component, {
-                ...populatedProps,
-                key: index,
-                onClick,
-                ref: item.focusable && ref,
-                role,
-                style,
-            });
+/**
+ * A react-window's List wrapper that instantiates the virtualized list and
+ * dynamically calculates the item height depending on the type
+ */
+class DropdownCoreVirtualized extends React.Component<Props, State> {
+    state = {
+        height: this.getHeight(),
+        width: null,
+    };
+
+    componentDidMount() {
+        // Wait for styles to be applied. This way, we can get a more precise
+        // value of the container dimensions.
+        setTimeout(() => {
```

Comment:
I think we definitely want to use the timing API for this! And I think we should be using `animationFrame` specifically as the styles should be laid out by that point.